### ISSUE TITLE
turf.intersect(geometry 1, geometry 2) gets stuck for endless time

### ISSUE THREAD

**Issue by usmanijaaz** [Turfjs/turf#2601]
- The version of Turf using: 6.5.0

I am using a 3rd party API to get geometry data based on a bounding box and that data consists of multi polygons with rings having hundreds of coordinates. Everything works fine in such cases, I recently came across an example where API response had multi polygon with 1 or 2 rings having 150k+ coordinates (geometry feature of some river or a forest) and that feature had 1000+ rings. 

Taking intersection of that multi polygon with my site polygon using `turf.intersect(incomingFeature, sitePolygon)` gets stuck and never complete its execution. 

**What I have done for now:** 
I have split incoming feature rings in two separate multi polygons, one with rings having coordinates greater than 10k and the other one with rings having coordinates less than 10k. turf.intersect seems to be working fine in this way, which seems weird. Why can't it process the multipolygon without splitting the rings ?

```
//Code snippet
//make polygon from incoming features
let incomingMultiPolygon = returnedFeatures[i].geometry.rings.filter(x => x.length <= 10000);
if (incomingMultiPolygon.length > 0)
  this.checkIntersection(incomingMultiPolygon, siteBoundary);

incomingMultiPolygon = returnedFeatures[i].geometry.rings.filter(x => x.length > 10000);
if (incomingMultiPolygon.length > 0) 
  this.checkIntersection(incomingMultiPolygon, siteBoundary);
```

//this.checkIntersection calls `turf.intersect(incomingMultiPolygon, siteBoundary)` and do something with the result.

**Comment by smallsaucepan** [comment]
Hi @usmanijaaz . Can you please attach geojson files of returnedFeatures and siteBoundary? Also, have you commented out other code in your checkIntersection function to eliminate other potential causes?

**Comment by usmanijaaz** [comment]
> Hi @usmanijaaz . Can you please attach geojson files of returnedFeatures and siteBoundary? Also, have you commented out other code in your checkIntersection function to eliminate other potential causes?

Hi @smallsaucepan , thanks for your response. I am attaching the siteBoundary file you requested and I'm also attaching a screenshot for your convenience to understand returnedFeatures data as I am not able to upload file due to its size. As you can see in the screenshot, there are multiple rings in geometry and 1st ring contains 155036 coordinates. 
 
And the first line of code in `checkIntersection` function is `if (turf.intersect(ourSiteBoundary, polygonFromReturnedFeatures))` ,  the code freezes here and never move past this line. 

![returnFeaturesScreenshot](https://github.com/Turfjs/turf/assets/72704740/236097d6-f8e8-49aa-aed7-54c34d18f107)
[siteboundary.json](https://github.com/Turfjs/turf/files/15198238/siteboundary.json)

**Comment by smallsaucepan** [comment]
Thanks @usmanijaaz . There will limitations in any system, especially when large datasets are involved. Ideally though we'd be able to handle those gracefully. Or at least add something in the documentation to set user expectations.

Do you have other options for providing returnedFeatures geojson as well e.g. dropbox? That's going to be key in reproducing and potentially narrowing down the cause.

**Comment by usmanijaaz** [comment]
> Thanks @usmanijaaz . There will limitations in any system, especially when large datasets are involved. Ideally though we'd be able to handle those gracefully. Or at least add something in the documentation to set user expectations.
> 
> Do you have other options for providing returnedFeatures geojson as well e.g. dropbox? That's going to be key in reproducing and potentially narrowing down the cause.


Hi @smallsaucepan, 
Sure thing, here is a link to a drive folder where I have uploaded the file for you: https://drive.google.com/file/d/1jhG28WNTWb6R9qznLfvTjYcSPyT-sdfc/view?usp=drive_link

**Comment by smallsaucepan** [comment]
Thanks @usmanijaaz . Investigated and found a safety check in the polygon-clipping library that is meant to stop infinite loops. The behaviour wasn't exactly what you were seeing (the code threw an error, rather than hanging), however it sounds too similar not to be related.

Taking the returnedFeatures json you provided, I loaded different combinations of the rings (big rings, small rings, all rings) and tried calling intersect. Big rings and small rings by themselves worked ok, but all rings together produced this error after a few seconds:

```
turf-issue-2601/node_modules/polygon-clipping/dist/polygon-clipping.cjs.js:1726
          throw new Error("Infinite loop when putting segment endpoints in a priority queue " + "(queue size too big).");
                ^
Error: Infinite loop when putting segment endpoints in a priority queue (queue size too big).
```

Looking further in to that, polygon-clipping has [some code that says](https://github.com/mfogel/polygon-clipping/blob/82b0ac2c35304bfea9cac24eabcbda32a808ee55/src/operation.js#L14) "if I need to run this loop more than a million times I'm probably in an infinite loop so abort". Turns out you can also hit this threshold with a big enough polygon!

polygon-clipping lets you override this ```env.POLYGON_CLIPPING_MAX_QUEUE_SIZE``` so from the command line you could try:

```
$ POLYGON_CLIPPING_MAX_QUEUE_SIZE=10000000 npx tsx ./test.ts
```
(that's 10 million)

Doing this allowed intersect to run through all the rings at once and produce the below polygon in about 5 seconds.

<img width="713" alt="Screenshot 2024-05-06 at 10 58 12 pm" src="https://github.com/Turfjs/turf/assets/5403628/ad99bdd3-3106-4b0a-b3c6-bb30c222a8d5">

Give that a try and see if it helps or changes the situation.

**Comment by usmanijaaz** [comment]
> Thanks @usmanijaaz . Investigated and found a safety check in the polygon-clipping library that is meant to stop infinite loops. The behaviour wasn't exactly what you were seeing (the code threw an error, rather than hanging), however it sounds too similar not to be related.
> 
> Taking the returnedFeatures json you provided, I loaded different combinations of the rings (big rings, small rings, all rings) and tried calling intersect. Big rings and small rings by themselves worked ok, but all rings together produced this error after a few seconds:
> 
> ```
> turf-issue-2601/node_modules/polygon-clipping/dist/polygon-clipping.cjs.js:1726
>           throw new Error("Infinite loop when putting segment endpoints in a priority queue " + "(queue size too big).");
>                 ^
> Error: Infinite loop when putting segment endpoints in a priority queue (queue size too big).
> ```
> 
> Looking further in to that, polygon-clipping has [some code that says](https://github.com/mfogel/polygon-clipping/blob/82b0ac2c35304bfea9cac24eabcbda32a808ee55/src/operation.js#L14) "if I need to run this loop more than a million times I'm probably in an infinite loop so abort". Turns out you can also hit this threshold with a big enough polygon!
> 
> polygon-clipping lets you override this `env.POLYGON_CLIPPING_MAX_QUEUE_SIZE` so from the command line you could try:
> 
> ```
> $ POLYGON_CLIPPING_MAX_QUEUE_SIZE=10000000 npx tsx ./test.ts
> ```
> 
> (that's 10 million)
> 
> Doing this allowed intersect to run through all the rings at once and produce the below polygon in about 5 seconds.
> 
> <img alt="Screenshot 2024-05-06 at 10 58 12 pm" width="713" src="https://private-user-images.githubusercontent.com/5403628/328186668-ad99bdd3-3106-4b0a-b3c6-bb30c222a8d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwMDEzMDMsIm5iZiI6MTcxNTAwMTAwMywicGF0aCI6Ii81NDAzNjI4LzMyODE4NjY2OC1hZDk5YmRkMy0zMTA2LTRiMGEtYjNjNi1iYjMwYzIyMmE4ZDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDZUMTMxMDAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWQzZDQ5Y2RkZDVkZTdiODlhZTBjNTIxNTkyZDgwYjBiOWU3OTNlOWZjMTgxOWM4OWRkMmMyMjJmN2FmYmM3NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2D9dYDD1iHOMoyfTg-EIE8S5pDPzaiOLCT1xnl1bGcQ">
> Give that a try and see if it helps or changes the situation.

Oh got it @smallsaucepan, thanks a lot for looking into this. I have currently deployed the solution where i split rings into two categories and process them separately and it is working fine for me.
However, I was wondering that I am using turf's npm package in my angular app, How can i set an environment variable for polygon_clipping library in my application.

**Comment by smallsaucepan** [comment]
Happy to help!

Hmm, setting it via Angular in the browser sounds a bit fiddly. Maybe take a look at this for some ideas - https://stackoverflow.com/a/46269080/3606644

Also note that if you increase that value and then get into the type of infinite loop the code is checking for, it will take 10x as long / use 10x as much memory before the failsafe activates.

Best of luck!

**Comment by smallsaucepan** [comment]
FYI @usmanijaaz we're almost finished upgrading to a different clipping library. Re-ran your data through that, and while it takes several seconds it produces an identical result to the `POLYGON_CLIPPING_MAX_QUEUE_SIZE=10000000` workaround.

![Image](https://github.com/user-attachments/assets/23be4455-e5b8-49fd-aa1a-b3d2dc8fd1ef)

This fix should hopefully be available soon in Turf v7.2.0 or 7.3.0
